Add JWTKit post (#22)
swift-on-server/articles: Sources/Articles/Documentation.docc/2025/jwt-kit/jwt-kit.md

> # Getting Started with Vapor's JWTKit 5: A Practical Introduction to JSON Web Tokens in Swift

Reviewer: This title should be shorter. Either the leading part of the trailing part. I think the latter part is more searchable for people coming into the Swift ecosystem.

Author: Updated!

> JWTKit is a library for working with JSON Web Tokens (JWT) in Swift. It provides a simple and easy-to-use API for creating, parsing, and validating JWTs. JWTs are a popular way to securely transmit information between parties, and JWTKit makes it easy to work with them in your Swift projects.

Reviewer: ```suggestion
JWTKit is a library for working with JSON Web Tokens (JWT) in Swift. It provides a simple and easy-to-use API for creating, parsing, and validating JWTs. JWTs are a popular way to authenticate information (claims) between parties. JWTKit makes it easy to work with them in your Swift projects.
```

Reviewer: It's not secure in the sense people would expect. JWEs do encrypt data, but a common JWT (JWS) only signs data. It's not "secure" in the sense of unreadable.

In this tutorial, you'll learn how to use JWTKit to create and validate JWTs in Swift. You'll start by installing JWTKit using Swift Package Manager, then you'll learn how to create a JWT and validate it using JWTKit's API. At the end of this tutorial, you'll be able to implement an authentication and authorisation flow in your Swift projects using JWTs.

## What is a JSON Web Token (JWT)?

> A JWT is a compact, URL-safe way to represent claims between two parties. The token is digitally signed, which makes it secure and trustworthy. JWTs are commonly used for authentication and authorisation in web applications, APIs, and microservices.

Reviewer: ```suggestion
A JWT is a compact, URL-safe way to represent claims between two parties. The token is digitally signed, which makes it secure and trustworthy. JWTs are commonly used for authentication and authorisation in (web) applications, APIs, and microservices.
```

Reviewer: Would like to make it clear this is common in iOS apps too (probably a large chunk of our audience)

> A JWT consists of three parts: a header, a payload, and a signature. The header contains metadata about the token, such as the type of token and the algorithm used to sign it. The payload contains the claims, which are statements about an entity (typically the user) and additional data. The signature is used to verify that the token is valid and has not been tampered with.

Reviewer: ```suggestion
A JWT consists of three parts:

1. A header, containing metadata about the token, such as the type of token and the algorithm used to sign it.
2. The payload, which contains the claims, which are statements about an entity (typically the user) and additional data.
3. And a signature. The signature is used to verify that the token is valid and has not been tampered with.
```

Reviewer: Breaking it down in bullet points makes it easier to scan

It usually looks like this:
```
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c
```

If you inspect the token on [jwt.io](https://jwt.io), you'll see that it consists of three parts separated by dots: the header, the payload, and the signature.

> Note: don't paste production tokens into jwt.io or any other online tool, as they could be intercepted and misused.

## Installing JWTKit

JWTKit is available as a Swift package, which makes it easy to install and use in your projects. To install it, add it as a dependency in your `Package.swift` file:

```swift
dependencies: [
    .package(url: "https://github.com/vapor/jwt-kit.git", from: "5.0.0")
]
```

Then, add JWTKit as a dependency to your target:

```swift
targets: [
    .target(name: "YourTarget", dependencies: [
        .product(name: "JWTKit", package: "jwt-kit")
    ])
]
```

## Configuration

> In JWTKit everything revolves around the ``JWTKeyCollection`` object. As the name suggests, it's a collection of keys that can be used to sign and verify JWTs. The declaration is simple, you can add this in your app's configuration code:

Reviewer: ```suggestion
In JWTKit everything revolves around the ``JWTKeyCollection`` object. This is a collection of keys that can be used to sign and verify JWTs. The declaration is simple, you can add this in your app's configuration code:
```

<!-- @Snippet(path: "site/Snippets/JWTKit", slice: key_collection_init) -->

```swift
let keyCollection = JWTKeyCollection()
```

> If you know the signing algorithm you want to use, adding keys to the collection is also straightforward:

Reviewer: ```suggestion
Adding keys to the collection is also straightforward:
```

Reviewer: Would you recommend HMAC SHA256? If so, let's stick to that and not confuse the reader. Maybe we can reference some reading materials on using the right signature. But if they're building their own backend, they likely have all the decision power and thus should use a sensible default.

> <!-- @Snippet(path: "site/Snippets/JWTKit", slice: key_collection_add_hmac) -->

Reviewer: ```suggestion
@Snippet(path: "site/Snippets/jwt-kit", slice: key_collection_add_hmac)
```

I believe your snippets didn't work because the filename isn't `JWTKit.swift` but `jwt-kit.swift` (note we need to omit the file extension)

```swift
await keyCollection.add(hmac: "secret", digestAlgorithm: .sha256)
```

> This snippet adds an HMAC (Hash-based Message Authentication Code) key to the collection. HMAC works like a digital signature for our JWT, but with a simpler approach than asymmetric algorithms:
> 1. First, the JWT's content (header and payload) is hashed using SHA-256
> 2. Then, this hash is combined with our secret key to create a unique signature
> 3. The same secret key must be used for both signing and verification (that's what makes it "symmetric")
> Think of it like a special seal on a letter - only someone with the same seal (our secret key) can verify if the letter is authentic. In this case, SHA-256 is our digest algorithm - it's responsible for creating that initial hash of our data. It's widely used because it provides a good balance of security and performance.

Reviewer: ```suggestion
This snippet adds an HMAC (Hash-based Message Authentication Code) key to the collection. HMAC works like a digital signature for our JWT.

1. First, the JWT's content (header and payload) is hashed using SHA-256
2. Then, this hash is combined with our secret key to create a unique signature
3. For HMAC, the same secret key must be used for both signing and verification (this is why we call this a "symmetric" algorithm)

### How signatures work

A JWT "signature" is a unique value that is derived from combining a key (the secret) with a payload through an algorithm (HMAC-SHA-256). This algorithm only produces that unique value when the same key is used on the same payload.

Because the secret is not shared with any other party, only your backend can produce that signature. Validation works by re-creating the same signature using the same payload and key. If using the payload and your secret key produce the exact same signature, this tells you that the payload is authentic and created by a holder of your key.
```

We should probably consider a crypto article

Reviewer: I think correctness matters here, for both people's understanding and correct use - so I'm suggesting the above revision.

Author: I agree that correctness matters so I removed most of the comment and added a link to a crypto article :) What do you say?

> Important: Since HMAC uses the same key for signing and verification, you must keep this secret key secure and never expose it in client-side code or public repositories. Use environment variables or a secure secrets manager to store it.

> > You might be wondering why the `add` method is marked with `await`. This is because the method is asynchronous due to `JWTKeyCollection` being an actor. You can read more about actors in [Joannis' article](https://swiftonserver.com/structured-concurrency-and-shared-state-in-swift/#Actors%20and%20Isolation).

Reviewer: ```suggestion
> You might be wondering why the `add` method is marked with `await`. This is because the method is asynchronous due to `JWTKeyCollection` being an actor. You can read more about actors in <doc:structured-concurrency-and-shared-state-in-swift>
```

Other than HMAC, JWTKit also supports **ECDSA**, **EdDSA** and **RSA** keys. These are asymmetric signing algorithms, which means the key used to sign the JWT is different from the key used to verify it. This article won't go into which algorithm is best as each one has its own use case, however it's important to note that RSA is the least recommended of the listed ones due to different reasons. That's why JWTKit hides RSA keys behind an `Insecure` namespace.

## Creating a JWT

Once you have a ``JWTKeyCollection`` object, you can use it to "create" a JWT. Creating a JWT means signing the payload with a key, in this case one from the collection. The payload is the data we want to transmit securely:

<!-- @Snippet(path: "site/Snippets/JWTKit", slice: payload_struct) -->

```swift
struct TestPayload: JWTPayload {
    var expiration: ExpirationClaim
    var issuer: IssuerClaim

    enum CodingKeys: String, CodingKey {
        case expiration = "exp"
        case issuer = "iss"
    }

    func verify(using key: some JWTAlgorithm) throws {
        try self.expiration.verifyNotExpired()
    }
}
```

> In this example, we define a ``TestPayload`` struct that conforms to the ``JWTPayload`` protocol. This protocol requires us to implement the `verify(using:)` method, which includes optional additional validation logic that can be performed when creating the JWT. In this case, we're verifying that the token has not expired.

Reviewer: ```suggestion
In this example, we define a ``TestPayload`` struct that conforms to the ``JWTPayload`` protocol. This protocol requires us to implement the `JWTPayload.verify(using:) method, which includes optional additional validation logic that can be performed when creating the JWT. In this case, we're verifying that the token has not expired.
```

The properties of the struct are the claims we want to include in the JWT. JWTKit provides a number of built-in claims, such as ``ExpirationClaim`` and ``IssuerClaim``, which are commonly used in JWTs. JWTKit supports the [seven registered claims](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1) defined in the JWT specification, but you can also define custom claims if needed. 

> > Note: in the example, `CodingKeys` are defined to map Swift property names to the JSON keys used in the JWT. This means that in the JWT the expiration claim will be named `exp` and the issuer claim will be named `iss`, as per the JWT specification.

Reviewer: ```suggestion
> Note: in the example, ``CodingKeys`` are defined to map Swift property names to the JSON keys used in the JWT. This means that in the JWT the expiration claim will be named `exp` and the issuer claim will be named `iss`, as per the JWT specification.
```

To create a JWT with this payload, you can create a new instance of the payload and use the key collection to sign it:

<!-- @Snippet(path: "site/Snippets/JWTKit", slice: jwt_sign) -->

```swift
let payload = TestPayload(
    expiration: .init(value: .init(timeIntervalSinceNow: 3600)),
    issuer: "myapp.com"
)

let token = try await keyCollection.sign(payload)
```

This will create a token that looks like the one we showed earlier. The token is now ready to be transmitted to the other party.

> ## Validating a JWT

Reviewer: ```suggestion
## Verifying a JWT
```

> Once you receive a JWT, you can use the key collection to validate it. Validation involves verifying the signature of the token and checking the claims in the payload:

Reviewer: ```suggestion
Once you receive a JWT, you can use the key collection to verify it. Verifying a JWT involves verifying the signature of the token and then checking the claims in the payload:
```

Should we rename **validate** to **verify** to stay consistent with your AI surface?

```swift
struct UserPayload: JWTPayload {
    let userID: UUID
    let expiration: ExpirationClaim
    let roles: [String]

    enum CodingKeys: String, CodingKey {
        case userID = "user_id"
        case expiration = "exp"
        case roles
    }

    func verify(using key: some JWTAlgorithm) throws {
        try expiration.verifyNotExpired()
        guard roles.contains("admin") else {
            throw JWTError.claimVerificationFailure(name: "roles", reason: "User is not an admin")
        }
    }

    init(from user: User) {
        self.userID = user.id
        self.expiration = .init(value: .init(timeIntervalSinceNow: 3600)) // Token expires in 1 hour
        self.roles = user.roles
    }
}

router.post("login") { req async throws -> Response in
    let user = User.find(username: req.body.username) // Find user by username, in a DB for example
    try user.verifyPassword(req.body.password)
    let payload = UserPayload(from: user)
    let token = try await keyCollection.sign(payload)
    return Response(status: .ok, body: token)
}
```

> The `UserPayload` struct represents the claims we want to include in the JWT. In this example, we include the user's ID, an expiration claim, and a list of roles. The `verify` method checks that the token has not expired and that the user has the "admin" role. The `init` method creates a new payload from a `User` object, which could be retrieved from a database, for example.

Reviewer: ```suggestion
The `UserPayload` struct represents the claims we want to include in the JWT. In this example, we include the user's ID, an expiration claim, and a list of roles. The ``JWTPayload/verify(using:)`` method checks that the token has not expired and that the user has the "admin" role. The `init` method creates a new payload from a `User` object, which could be retrieved from a database, for example.
```
